src/connector/src/sink/redis.rs

```diff
@@ -12,39 +12,302 @@
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RedisSink")
+            .field("config", &self.config)
+            .field("schema", &self.schema)
+            .field("client", &self.client)
+            .field("batch_id", &self.batch_id)
+            .field("epoch", &self.epoch)
+            .field("update_cache", &self.update_cache)
+            .field("pk_indices", &self.pk_indices)
+            .finish()
+    }
+}
 
 impl RedisSink {
-    pub fn new(_cfg: RedisConfig, _schema: Schema) -> Result<Self> {
-        todo!()
+    pub fn new(cfg: RedisConfig, schema: Schema, pk_indices: Vec<usize>) -> Result<Self> {
+        let client = redis::Client::open(cfg.endpoint.clone())?;
+        let conn = client.get_connection()?;
+        let pipe = redis::pipe();
+        Ok(Self {
+            config: cfg,
+            schema,
+            client: Some(client),
+            conn: RedisConnectionImpl::Redis(conn),
+            pipe,
+            batch_id: 0,
+            epoch: 0,
+            update_cache: Vec::new(),
+            pk_indices,
+        })
+    }
+
+    #[cfg(test)]
+    fn for_test(schema: Schema, pk_indices: Vec<usize>) -> Result<Self> {
+        let pipe = redis::pipe();
+        Ok(Self {
+            config: RedisConfig {
+                endpoint: "".parse().unwrap(),
+            },
+            schema,
+            client: None,
+            conn: RedisConnectionImpl::Mock(MockConnection {}),
+            pipe,
+            batch_id: 0,
+            epoch: 0,
+            update_cache: Vec::new(),
+            pk_indices,
+        })
+    }
+
+    pub fn make_redis_key(row: RowRef<'_>, pk_indices: &[usize]) -> String {
+        pk_indices
+            .iter()
+            .map(|i| row.value_at(*i).to_text())
+            .join(":")
+    }
+}
+
+pub struct MockConnection {}
+
+impl ConnectionLike for MockConnection {
+    fn req_packed_command(&mut self, _cmd: &[u8]) -> RedisResult<Value> {
+        Ok(Value::Okay)
+    }
+
+    fn req_packed_commands(
+        &mut self,
+        _cmd: &[u8],
+        _offset: usize,
+        _count: usize,
+    ) -> RedisResult<Vec<Value>> {
+        Ok(Vec::new())
+    }
+
+    fn get_db(&self) -> i64 {
+        0
+    }
+
+    fn check_connection(&mut self) -> bool {
+        true
+    }
+
+    fn is_open(&self) -> bool {
+        true
     }
 }
 
 #[async_trait]
 impl Sink for RedisSink {
-    async fn write_batch(&mut self, _chunk: StreamChunk) -> Result<()> {
-        todo!();
+    async fn write_batch(&mut self, chunk: StreamChunk) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            let key = Self::make_redis_key(row, &self.pk_indices);
+            match op {
+                Op::Insert => {
+                    // current format: key=`pk1[:pk2:pk3:...]` value="[v1,v2,...]"
+                    self.pipe.set(
+                        key,
+                        format!("[{}]", row.values().map(|v| v.to_text()).join(",")),
```

Comment:
just `json!` then `to_string` is ok

Reply:
I tried but it seems `RowRef` and `Row` has not implemented `Serialize`, we still have to do it column-wise?

Reply:
Just leave it for now.

I think letting users define the key format and value format is a good practice. Looking forward to it